atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStore.java

```diff
@@ -88,15 +86,13 @@ private List<LockWatchEvent> retentionEventsInternal(int numToRetention, Sequenc
         List<LockWatchEvent> events = new ArrayList<>(numToRetention);
 
         // The correctness of this depends upon eventMap's entrySet returning entries in ascending sorted order.
-        List<Map.Entry<Sequence, LockWatchEvent>> eventsToClear = eventMap.entrySet().stream()
+        eventMap.entrySet().stream()
+                .takeWhile(entry -> entry.getKey().value() < maxVersion.value())
                 .limit(numToRetention)
-                .filter(entry -> entry.getKey().value() < maxVersion.value())
-                .collect(Collectors.toList());
```

Comment:
Previously this would always traverse at least `numRetention` entries even if they would not satisfy the filter. We now only take any that satisfy the filter first (and eagerly terminate since the source is ordered), then limit to max `numToRetention`, as well as avoid the intermediate collect to list.